File: cpp_solutions/sequence/30_Substring_With_Concatenation_of_All_Words.cpp
/*

You are given a string, s, and a list of words, words, that are all of the same length. Find all starting indices of substring(s) in s that is a concatenation of each word in words exactly once and without any intervening characters.

For example, given:
s: "barfoothefoobarman"
words: ["foo", "bar"]

You should return the indices: [0,9].
(order does not matter).

*/

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> rst;
        if(words.size() == 0) return rst;
        
        unordered_map<string, int> dict;
        for(int i=0; i< words.size(); i++)
        {
            if(dict.find(words[i]) == dict.end())
                dict[words[i]] = 1;
            else
                dict[words[i]]++;
        }
        
        int wSize = words[0].length();
        for(int i=0; i< wSize; i++)
        {
            int wCnt = 0;
            unordered_map<string, int> slideWin;
            for(int j=i; j + wSize<= s.length(); j+=wSize)
            {
                string curWord = s.substr(j, wSize);
                if(dict.find(curWord)==dict.end())
                {
                    slideWin.clear();
                    wCnt=0;
                }
                else
                {
                    wCnt++;
                    if(slideWin.find(curWord) == slideWin.end())
                    {
                        slideWin[curWord] = 1;
                    }
                    else
                    {
                        slideWin[curWord]++;
                    }
                    
                    while(dict[curWord]< slideWin[curWord])
                    {
                        string rmWord = s.substr(j-(wCnt-1)*wSize, wSize);
                        slideWin[rmWord]--;
                        wCnt--;
                    }
                }
                
                if(wCnt == words.size())
                {
                    rst.push_back(j - (wCnt-1)*wSize);
                }
            }
        }
        
        return rst;
    }
};

File: python_solutions/Frequency/30.Substring_with_Concatenation_of_All_Words.py
class Solution:
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        ret = []
        if len(words) == 0:
            return ret
        
        m = collections.defaultdict(int)
        for w in words:
            m[w] += 1
        
        wSize = len(words[0])
        for i in range(wSize):
            win = collections.defaultdict(int)
            wCnt = 0
            for j in range(i, len(s)-wSize+1, wSize):
                curWord = s[j:j+wSize]
                if curWord not in m:
                    win.clear()
                    wCnt = 0
                else:
                    win[curWord] += 1
                    wCnt += 1
                    
                    # --> right
                    while win[curWord] > m[curWord]:
                        idx = j-(wCnt-1)*wSize  # rightest word
                        rmvWord = s[idx :  idx+wSize]
                        win[rmvWord] -= 1
                        wCnt -= 1
                if wCnt == len(words):
                    ret += j-(wCnt-1)*wSize,
            
        return ret
                        
        